File: leetcode/3200-3300/3290. 最高乘法得分.cpp
//
// Created by 123 on 25-5-23.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;

class Solution {
public:
  long long maxScore(vector<int>& a, vector<int>& b) {
    int n = b.size();
    vector<vector<long long>> dp(5, vector<long long>(n + 1));
    for (int i = 1; i <= 4; i++) {
      dp[i][0] = LONG_LONG_MIN / 2;
      for (int j = 1; j <= n; j++) {
        dp[i][j] = max(dp[i][j - 1], dp[i - 1][j - 1] + 1ll * a[i - 1] * b[j - 1]);
      }
    }
    return dp[4][n];
  }
};

class Solution2 {
public:
  long long maxScore(vector<int>& a, vector<int>& b) {
    int n = b.size();
    vector<array<long long, 5>> f(n + 1);
    for (int j = 1; j < 5; j++) {
      f[0][j] = LLONG_MIN / 2;
    }
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < 4; j++) {
        f[i + 1][j + 1] = max(f[i][j + 1], f[i][j] + (long long) a[j] * b[i]);
      }
    }
    return f[n][4];
  }
};

File: leetcode/3200-3300/3290. 最高乘法得分.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

MOD = int(1e9 + 7)
DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]


class Solution:
    def maxScore(self, a: List[int], b: List[int]) -> int:
        n = len(b)
        dp = [[-inf] * (n + 1) for _ in range(5)]
        for i in range(n + 1):
            dp[0][i] = 0

        for i in range(1, 5):
            for j in range(1, n + 1):
                dp[i][j] = max(dp[i][j - 1], dp[i - 1][j - 1] + a[i - 1] * b[j - 1])
        return dp[4][n]
